clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -241,7 +241,7 @@ public void run() {
         // okay we stopped accepting requests but there may still be
         // requests in the accumulator or waiting for acknowledgment,
         // wait until these are completed.
-        while (!forceClose && (this.accumulator.hasUndrained() || this.client.inFlightRequestCount() > 0)) {
+        while (!forceClose && (this.accumulator.hasUndrained() || this.client.inFlightRequestCount() > 0 || transactionManager.hasOngoingTransaction())) {
```

Comment:
A couple things to consider:
1. If close() is called and a transaction has not been committed or aborted, should we abort it explicitly?
2. I mentioned in the JIRA that the thread blocking on `commitTransaction()` may be stuck if we shutdown the `Sender` before the future has been notified. That seems to still be a problem as far as I can tell. Maybe we should add a `TransactionManager.close()` which does some cleanup.

Reply:
Tried to explore the answers, I think the following:

1. In case of `close()` it depends. 
a.1) If the producer is only idempotent and not transactional then I believe we can abort the current transaction explicitly as some parts of it may not have made it even to the accumulator, therefore we won't be able to commit it. It's better than leaving it hanging around until the transaction timeout. 
a.2) On `forceClose` we should leave it as it is as the coordinator will abort the transaction after the timeout. I think aborting in this case doesn't make sense as that way `forceClose` would lose its purpose.
b) If the producer is transactional, then we wouldn't want to close it explicitly as it may be committed in the next session, therefore on that front we probably have nothing to do.
2. Yeah, I see the problem now, I'll work on it.

Reply:
What did you mean by aborting it explicitly? As I understand incomplete transactions may be continued in a next application session, therefore on close we wouldn't want to abort the transaction itself, just the current queue as per my previous answer.

Reply:
Ok, reading KIP-98 again I think you were referring to this point:

> 2. Transaction recovery across application sessions. If an application instance dies, the next instance can be guaranteed that any unfinished transactions have been completed (whether aborted or committed), leaving the new instance in a clean state prior to resuming work. 

So in this case it is fair that we should abort the transaction if we can't commit it. But what about trying to finish the commit? As far as I see in this case it would be possible as `initiateClose()` has been called, the commit request has been enqueued in TransactionManager but the sender could be sleeping as no one called `wakeup()` on it. Therefore we could just do a check on the enqueued requests, flush them and then close TransactionManager.
I could imagine a case where we don't block on `commitTransaction()` but `sendOffsetsToTransaction()` where we could send out an abort for instance.

Reply:
To clarify, I was suggesting that we can abort a pending transaction only if `close()` is called before the user has attempted to commit.